spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -193,47 +193,438 @@ For example, calling an `@Outgoing` annotated method from user code will not pos
+Note that `@Incoming("channel") void method(Message<I> msg)` is not allowed as message acknowledgement is asynchronous.
+
+|
+[source,java]
+----
+@Incoming("channel") 
+CompletionStage<?> method(I payload)
+----
+| Consumes the _payload_ asynchronously
+| This method is called for every `Message<I>` instance transiting on the channel `channel`. 
+The payload is automatically extracted from the inflight messages using `Message.getPayload()`.
+The user method is never called concurrently. The reactive messaging implementation must wait until the completion of the previously returned `CompletionStage` before calling the method again with the next _payload_.
+
+|===
+	
+===== Methods processing data
+
+[cols="2a,1,1",options="header"]
+|===
+|Signature
+|Behavior
+|Invocation
+
+| 
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out")
+Processor<Message<I>, Message<O>> method()
+----
+| Returns a Reactive Streams processor consuming incoming `Message` instances and produces `Message` instances.
+| This method is called once; at assembly time.
+
+| 
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Processor<I, O> method();
+----
+| Returns a Reactive Streams processor consuming incoming _payload_ instances and produces _payload_ instances.
+| This method is called once; at assembly time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+ProcessorBuilder<Message<I>, Message<O>> method();
+----
+| Returns a `ProcessorBuilder` consuming incoming `Message` instances and produces `Message` instances.
+| This method is called once; at assembly time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+ProcessorBuilder<I, O> method();
+----
+| Returns a Reactive Streams processor that consuming incoming _payload_ instances and produces _payload_ instances.
+| This method is called once; at assembly time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Publisher<Message<O>> method(Message<I> msg)
+----
+| Returns a Reactive Streams `Publisher` for each incoming `Message`.
+The returned `Publisher` can be empty or emits multiple `Message` instances.
+If the returned `Publisher` emits several elements, these elements are _flatten_ in the outgoing stream as a concatenation of elements.
+| This method is called for every incoming message. Never concurrently, so the stream returned previously must complete before having the method invoked with the next incoming `Message`.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Publisher<O> method(I payload)
+----
+| Returns a Reactive Streams `Publisher` for each incoming _payload_.
+The returned `Publisher` can be empty or emits multiple _payload_ instances.
+If the returned `Publisher` emits several elements, these elements are _flatten_ in the outgoing stream as a concatenation of elements.
```

Comment:
What function is used to do the flattening?

Reply:
The used strategy is: `concatenation of the elements`. 
I'm going to add "The _flattening_ follows the same semantics as the `flatMap` operator from the MicroProfile Reactive Streams specification."